### PR TITLE
Add integration test for multiple audience in structured authn

### PR DESCRIPTION
**Pull request by aramase**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
- Follow-up to https://github.com/kubernetes/kubernetes/pull/123165. Addresses https://github.com/kubernetes/kubernetes/pull/123165#pullrequestreview-1880856268.
- Added integration tests and unit tests for nuanced audience validation using claim validation rules.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go [hunk 1]
```diff
@@ -1585,6 +1618,45 @@ func TestToken(t *testing.T) {
 			}`, valid.Unix()),
 			wantErr: `oidc: verify token: oidc: expected audience in ["my-client" "random-client"] got ["not-my-client"]`,
 		},
+		{
+			name: "nuanced audience validation using claim validation rules",
```
Comment (enj) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:1622]: Can you also add this same test but with a token that fails?

Comment (aramase) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:1622]: Done.

**Comment by aramase** [comment]
/assign liggitt

(for the API edits)

**Comment by aramase** [comment]
https://github.com/kubernetes/kubernetes/pull/123282 merged, so had to rebase and update integration test with the generics.

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go [hunk 2]
```diff
@@ -219,6 +219,7 @@ func validateClaimValidationRules(compiler authenticationcel.Compiler, celMapper
 
 			compilationResult, err := compileClaimsCELExpression(compiler, &authenticationcel.ClaimValidationCondition{
 				Expression: rule.Expression,
+				Message:    rule.Message,
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:222]: huh... was this being missing affecting validation / compilation?

**Review by liggitt — COMMENTED** [review]
/lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go (hunk 2 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:222]: I _think_ it just made the error message not useful.

**Pull request merged**